include/picongpu/simulationControl/AdjustDomainSize.hpp

```diff
@@ -0,0 +1,363 @@
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <pmacc/dimensions/DataSpace.hpp>
+
+#include <pmacc/Environment.hpp>
+#include <pmacc/mpi/reduceMethods/Reduce.hpp>
+#include <pmacc/mpi/MPIReduce.hpp>
+#include <pmacc/nvidia/functors/Add.hpp>
+#include <pmacc/nvidia/functors/Max.hpp>
+#include <pmacc/nvidia/functors/Min.hpp>
+#include <pmacc/dimensions/DataSpace.hpp>
+#include <pmacc/mpi/GetMPI_StructAsArray.hpp>
+
+
+namespace picongpu
+{
+    /** adjust domain sizes
+     *
+     * Extend the local offset, the local and global domain size to fulfill all
+     * PIConGPU/PMacc conditions domain.
+     */
+    class AdjustDomainSize
+    {
+
+    public:
+        /** constructor
+         *
+         * @param numDevices number of devices [per dimension]
+         * @param mpiPosition the position of the local device [per dimension]
+         * @param isPeriodic if the outer simulation boundaries are periodic [per dimension]
+         *                   1 is meaning periodic else 0
+         * @param movingWindowEnabled if moving window is enabled
+         */
+        AdjustDomainSize(
+            DataSpace< simDim > const & numDevices,
+            DataSpace< simDim > const & mpiPosition,
+            DataSpace< simDim > const & isPeriodic,
+            bool const movingWindowEnabled
+        ) :
+            m_numDevices( numDevices),
+            m_mpiPosition( mpiPosition ),
+            m_isPeriodic( isPeriodic),
+            m_movingWindowEnabled( movingWindowEnabled ),
+            m_isMaster( mpiPosition == DataSpace< simDim >::create( 0 ) )
+        {
+        }
+
+        /** adjust the correct grid size
+         *
+         * @param[in,out] globalDomainSize size of the global volume [in cells]
+         * @param[in,out] localDomainSize size of the local volume [in cells]
+         * @param[out] localDomainOffset local offset [in cells] relative to the origin of the global domain
+         */
+        void operator()(
+            DataSpace< simDim > & globalDomainSize,
+            DataSpace< simDim > & localDomainSize,
+            DataSpace< simDim > & localDomainOffset
+        )
+        {
+            m_globalDomainSize = globalDomainSize;
+            m_localDomainSize = localDomainSize;
+
+            for( uint32_t d = 0; d < simDim; ++d )
+            {
+                multipleOfSuperCell( d );
+                minThreeSuperCells( d );
+                greaterEqualThanAbsorber( d );
+                deriveGlobalDomainSize( d );
+                updateLocalDomainOffset( d );
+            }
+
+            // write results back
```

Comment:
Question: is it not possible have smth like `sizeChanged = (m_globalDomainSize != m_globalDomainSize) || (m_localDomainSize !=  localDomainSize) || (m_localDomainOffset != localDomainOffset )` here? (Instead of having it in different places throughout this class)